src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -185,14 +182,108 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
   return bestFoundGain;
 }
 
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::Row<double>& labels,
+    const size_t numClasses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimumLeafSize * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::Row<double> sortedLabels(labels.n_elem);
+  arma::rowvec sortedWeights;
+  for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+    sortedLabels[i] = labels[sortedIndices[i]];
```

Comment:
```suggestion
sortedLabels = labels(sortedIndices)
```
If you do decide to go with it, you'd have to enforce this pattern for the entire module for consistency reasons. It's your choice. IMO  arma > STL > raw for.

Reply:
Actually, at this point in time, I am not much concerned about these micro-changes. This is a very big refactoring and I don't want to get distracted by these things. My main focus right now is to do the refactoring correctly, so that I do not introduce any new bug. We can definitely deal with these small issues later.

Reply:
Good. Nice work btw, would be cool to have this merged!